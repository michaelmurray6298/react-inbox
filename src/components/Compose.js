import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { submitForm } from "../actions";
import {Redirect} from "react-router-dom"

class Compose extends Component {
	constructor(props) {
		super(props);
		this.state = {
			fireRedirect: false
		}
		this.handleForm = this.handleForm.bind(this);
	}
	handleForm(e) {
		e.preventDefault();
		this.props.submitForm({
			subjectValue: e.target.subject.value,
			bodyValue: e.target.body.value
		});
		this.setState({ fireRedirect: true })

	}

	render() {
		const { fireRedirect } = this.state
		return (
			<form className="form-horizontal well" onSubmit={this.handleForm}>
				<div className="form-group">
					<div className="col-sm-8 col-sm-offset-2">
						<h4>Compose Message</h4>
					</div>
				</div>
				<div className="form-group">
					<label htmlFor="subject" className="col-sm-2 control-label">
						Subject
					</label>
					<div className="col-sm-8">
						<input
							defaultValue="Subject"
							type="text"
							className="form-control"
							id="subject"
							placeholder="Enter a subject"
							name="subject"
						/>
					</div>
				</div>
				<div className="form-group">
					<label htmlFor="body" className="col-sm-2 control-label">
						Body
					</label>
					<div className="col-sm-8">
						<textarea
							defaultValue="Body"
							name="body"
							id="body"
							className="form-control"
						/>
					</div>
				</div>
				<div className="form-group">
					<div className="col-sm-8 col-sm-offset-2">
						<input type="submit" value="Send" className="btn btn-primary" />
					</div>

				{fireRedirect && (
          <Redirect to='/'/>
        )}
			</div>
		</form>
		);
	}
}

const mapStateToProps = state => {
	const messages = state.messages;
	return {
		messages
	};
};

const mapDispatchToProps = dispatch =>
	bindActionCreators(
		{
			submitForm
		},
		dispatch
	);

export default connect(mapStateToProps, mapDispatchToProps)(Compose);
