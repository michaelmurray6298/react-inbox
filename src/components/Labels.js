import React, { PureComponent } from 'react';

class Labels extends PureComponent {
  render() {
    return (
      <span className="label label-warning">
        {this.props.label}
      </span>
    );
  }
}

export default Labels;
