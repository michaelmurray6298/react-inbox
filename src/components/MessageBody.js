import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { fetchMessagesById } from '../actions';
// import { Link, BrowserRouter as Router, Route  } from "react-router-dom"


class MessageBody extends Component {
  componentDidMount() {
    this.props.fetchMessagesById(this.props.location);
  }
  render() {
    return (
      <div className="row message-body">
        <div className="col-xs-11 col-xs-offset-1">
          {this.props.messageById.body}
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchMessagesById,
    },
    dispatch,
  );
const mapStateToProps = (state) => {
  const messages = state.messages;
  const messageById = state.messagesById;
  return {
    messages,
    messageById,
  };
};


export default connect(mapStateToProps, mapDispatchToProps)(MessageBody);
