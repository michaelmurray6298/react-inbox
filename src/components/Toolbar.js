import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import {
	updateReadMessages,
	updateUnreadMessages,
	updateAll,
	updateLabelState
} from "../actions";

class Toolbar extends Component {
	calculateSelected() {
		let selectAll = "minus-";
		let selectedMsgs = this.props.messages.filter(msg => msg.selected === true);

		if (!selectedMsgs.length) {
			selectAll = "";
		} else if (selectedMsgs.length === this.props.messages.length) {
			selectAll = "check-";
		}

		return selectAll;
	}
	selectAllMessages() {
		if (this.calculateSelected() !== "check-") {
			this.props.updateAll({ selected: true });
		} else {
			this.props.updateAll({ selected: false });
		}
	}
	selectedMessagesCount() {
		let selectedMessagesCount = this.props.messages.filter(
			message => message.selected
		);
		return selectedMessagesCount.length;
	}
	unreadMessageCount() {
		let unreadMsgs = this.props.messages.filter(msg => msg.read === false);
		return unreadMsgs.length;
	}

	addLabel(event) {
		let newLabel = event.target.value;
		if (newLabel === "Apply label") {
			return;
		}
		this.props.updateLabelState(newLabel, "add");
	}

	deleteLabel(event) {
		let selectedLabel = event.target.value;
		if (selectedLabel === "Remove label") {
			return;
		}
		this.props.updateLabelState(selectedLabel);
	}

	deleteMessages() {
		this.props.updateRemovedMessages();
	}

	render() {
		const disabled = this.selectedMessagesCount() > 0 ? "" : "disabled";
		return (
			<div className="row toolbar">
				<div className="col-md-12">
					<p className="pull-right">
						<span className="badge badge">{this.unreadMessageCount()}</span>
						unread messages
					</p>
					<a className="btn btn-danger" onClick={() => this.props.compose()}>
						<i className="fa fa-plus" />
					</a>
					<button
						className="btn btn-default"
						onClick={() => this.selectAllMessages()}>
						<i className={`fa fa-${this.calculateSelected()}square-o`} />
					</button>

					<button
						className="btn btn-default"
						disabled={disabled}
						onClick={() => this.props.updateReadMessages()}>
						Mark As Read
					</button>

					<button
						className="btn btn-default"
						disabled={disabled}
						onClick={() => this.props.updateUnreadMessages()}>
						Mark As Unread
					</button>

					<select
						className="form-control label-select"
						disabled={disabled}
						onChange={event => this.addLabel(event)}>
						<option>Apply label</option>
						<option value="dev">dev</option>
						<option value="personal">personal</option>
						<option value="gschool">gschool</option>
					</select>

					<select
						className="form-control label-select"
						disabled={disabled}
						onChange={event => this.deleteLabel(event)}>
						<option>Remove label</option>
						<option value="dev">dev</option>
						<option value="personal">personal</option>
						<option value="gschool">gschool</option>
					</select>

					<button
						className="btn btn-default"
						disabled={disabled}
						onClick={() => this.deleteMessages()}>
						<i className="fa fa-trash-o" />
					</button>
				</div>
			</div>
		);
	}
}

const mapStateToProps = state => {
	const messages = state.messages;
	return {
		messages
	};
};

const mapDispatchToProps = dispatch =>
	bindActionCreators(
		{
			updateReadMessages,
			updateUnreadMessages,
			updateAll,
			updateLabelState
		},
		dispatch
	);

export default connect(mapStateToProps, mapDispatchToProps)(Toolbar);
