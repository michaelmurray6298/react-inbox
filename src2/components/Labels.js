import React, { Component } from "react";

class Labels extends Component {
	render() {
		return (
			<span className="label label-warning">
				{this.props.label}
			</span>
		);
	}
}

export default Labels;
