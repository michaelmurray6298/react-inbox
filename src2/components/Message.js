import React, { Component } from "react";
import Label from "./Labels.js";

class Message extends Component {
	toggleStarred() {
		let isStarred = !this.props.starred;
		fetch(`http://localhost:8181/api/messages`, {
			method: "PATCH",
			headers: {
				Accept: "application/json",
				"Content-Type": "application/json"
			},
			body: JSON.stringify({
				messageIds: [this.props.id],
				command: "star",
				star: isStarred
			})
		}).then(() => {
			this.props.updateState(this.props.id, { starred: isStarred });
		});
	}

	toggleChecked() {
		let isSelected = !this.props.selected;
		this.props.updateState(this.props.id, { selected: isSelected });
	}

	render() {
		const read = this.props.read ? "read" : "unread";
		const selected = this.props.selected ? "selected" : "";
		const checked = this.props.selected ? "checked" : "";
		const starred = this.props.starred ? "fa-star" : "fa-star-o";
		return (
			<div className={`row message ${read} ${selected}`}>
				<div className="col-xs-1">
					<div className="row">
						<div className="col-xs-2">
							<input
								type="checkbox"
								checked={`${checked}`}
								onChange={() => this.toggleChecked()}
							/>
						</div>
						<div className="col-xs-2">
							<i
								className={`star fa ${starred}`}
								onClick={() => {
									this.toggleStarred();
								}}
							/>
						</div>
					</div>
				</div>
				<div className="col-xs-11">
					{this.props.labels.map(label =>
						<Label label={label} key={label.id} />
					)}
					<span>
						{this.props.subject}
					</span>
				</div>
			</div>
		);
	}
}

export default Message;
