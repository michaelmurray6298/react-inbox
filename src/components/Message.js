import React, { PureComponent } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link, Route } from 'react-router-dom';
import { toggleStarred, toggleSelected, updateReadMessagesById } from '../actions';
import MessageBody from './MessageBody';
import Label from './Labels';

const MessageBodyRoute = props => (
  <MessageBody location={props.location.pathname} match={props.match} />
);

export class Message extends PureComponent {
  render() {
    const read = this.props.read ? 'read' : 'unread';
    const selected = this.props.selected ? 'selected' : "";
    const checked = this.props.selected ? 'checked' : "";
    const starred = this.props.starred ? 'fa-star' : 'fa-star-o';
    return (
        <div id="read-value" className={`row message ${read} ${selected}`}>
          <div className="col-xs-1">
            <div className="row">
              <div className="col-xs-2">
                <input
									id="select-button"
                  type="checkbox"
                  checked={`${checked}`}
                  onChange={() => this.props.toggleSelected(this.props.id)}
                />
              </div>
              <div className="col-xs-2">
                <i
									id="star-button"
                  className={`star fa ${starred}`}
                  onClick={() => {
                    this.props.toggleStarred(this.props.id, this.props.starred);
                  }}
                />
              </div>
            </div>
          </div>
          <div className="col-xs-11">
            {this.props.labels.map(label =>
              <Label label={label} key={this.props.id} />,
            )}
            <Link onClick={() => this.props.updateReadMessagesById(this.props.id)} to={`/messages/${this.props.id}`}>{this.props.subject} </Link>

            <Route exact path={`/messages/${this.props.id}`} component={MessageBodyRoute} />
          </div>
        </div>
    );
  }
}
export const mapStateToProps = (state) => {
  const messages = state.messages;
  return {
    messages,
  };
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      toggleStarred,
      toggleSelected,
      updateReadMessagesById,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(Message);
