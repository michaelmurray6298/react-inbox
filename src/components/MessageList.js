import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { fetchMessages } from '../actions';
import Message from './Message.js';

class MessageList extends Component {
  componentDidMount() {
    this.props.fetchMessages();
  }

  render() {
    const { messages } = this.props;

    return (
      <div className="container">
        {messages.map(message =>
          (<Message
            key={message.id}
            id={message.id}
            labels={message.labels}
            selected={message.selected}
            starred={message.starred}
            read={message.read}
            subject={message.subject}
            location={this.props.location}
          />),
        )}
      </div>
    );
  }
}
const mapStateToProps = (state) => {
  const messages = state.messages;
  const messageById = state.messagesById;
  return {
    messages,
    messageById,
  };
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchMessages,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(MessageList);
