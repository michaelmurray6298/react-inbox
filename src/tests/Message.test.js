import React from 'react';
import { shallow, render, mount } from 'enzyme';
import toJson, { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import { Message, mapStateToProps } from '../components/Message';

function setup() {
  const props = {
    toggleStarred: jest.fn(),
    toggleSelected: jest.fn(),
    selected: false,
    read: false,
    starred: false,
    id: 1,
    labels: [],
    subject: 'Hi',
  };

  const enzymeWrapper = mount(<MemoryRouter>{<Message {...props} />}</MemoryRouter>);
  return {
    props,
    enzymeWrapper,
  };
}

describe('Message', () => {
  it('should change classnames when the message is starred vs unstarred', () => {
    const { enzymeWrapper, props } = setup();
    enzymeWrapper.find('#select-button').simulate('change');
    expect(props.toggleSelected).toHaveBeenCalled();
  });

  it('should change classnames when the message is selected vs unselected', () => {
    const { enzymeWrapper, props } = setup();
    enzymeWrapper.find('#star-button').simulate('click');
    expect(props.toggleStarred).toHaveBeenCalled();
  });

  it('should change classnames when the message is read vs unread', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find('#read-value').props().className).toEqual('row message unread ');
  });

  it('should render', () => {
    const component = shallow(<Message
      selected={false}
      read={false}
      starred={false}
      id={1}
      labels={[]}
      subject={'Hi'}
      location={'/message/1'}
    />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('mapStateToProps', () => {
    const state = {
      messages: [
        {
          id: 1,
          labels: [],
          read: false,
          selected: false,
          starred: false,
          subject: 'Hi',
        },
      ],
    };

    const expectedResult = {
      messages: [
        {
          id: 1,
          labels: [],
          read: false,
          selected: false,
          starred: false,
          subject: 'Hi',
        },
      ],
    };

    expect(mapStateToProps(state)).toEqual(expectedResult);
  });
});
