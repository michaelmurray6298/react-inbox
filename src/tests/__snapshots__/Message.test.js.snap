// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render 1`] = `
<div
  className="row message unread "
  id="read-value"
>
  <div
    className="col-xs-1"
  >
    <div
      className="row"
    >
      <div
        className="col-xs-2"
      >
        <input
          checked=""
          id="select-button"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <div
        className="col-xs-2"
      >
        <i
          className="star fa fa-star-o"
          id="star-button"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="col-xs-11"
  >
    <Link
      onClick={[Function]}
      replace={false}
      to="/messages/1"
    >
      Hi
       
    </Link>
    <Route
      component={[Function]}
      exact={true}
      path="/messages/1"
    />
  </div>
</div>
`;
